api/agent.go

```diff
@@ -1237,15 +1238,15 @@ func (a *Agent) updateTokenOnce(target, token string, q *WriteOptions) (*WriteMe
 	if err != nil {
 		return nil, 0, err
 	}
-	defer closeResponseBody(resp)
 
 	wm := &WriteMeta{RequestTime: rtt}
 
-	if resp.StatusCode != 200 {
-		var buf bytes.Buffer
-		io.Copy(&buf, resp.Body)
-		return wm, resp.StatusCode, fmt.Errorf("Unexpected response code: %d (%s)", resp.StatusCode, buf.Bytes())
+	_, _, err = requireOK(rtt, resp, err)
+
+	if err != nil {
+		return wm, resp.StatusCode, err
```

Comment:
I'm a little confused about the order in which `requireOK` is called here - should this be handled close to L1238 right after `doRequest(r)`?

Reply:
Yes I think so -- thanks for spotting that!

PS: I think when I chose not to I forgot we actually check the err from `doRequest()` inside the require calls!

https://github.com/hashicorp/consul/blob/5c37819d0957f54242d46650e74756346e4e1b2c/api/api.go#L1107-L1111

Reply:
Let me follow up here to explain the old odd/ smelly ordering!

So in the snippet above, if there is an error during the `doRequest()`, and we swallow the `resp` object and throw it back as `nil`. 

So there was not an easy way to combine the first err check, where we want to return 0 for some reason to the callee of `udpateTokenOnce()`, with the synthetic error checks.

As of now, I tried to re-write the snippet of code to make more sense -- hopefully it's clearer now but please let me know if it's not.